academy/0.0-B9lab-Blockchains/5_crypto.md

```diff
@@ -0,0 +1,180 @@
+
+#### Encrypt and decrypt
+
+Alice wants to send a message to Bob, and for Bob's eyes only:
+
+* Bob gives Alice his public key
+* Alice uses Bob's public key to encrypt the message
+* Alice sends Bob the encrypted message
+* Bob decrypts the message with his private key
+
+![](images/00_11_rsa_keys_v1.png)
+
+Example:
+
+```bash
+// Encrypt file
+$ openssl pkeyutl -encrypt -pubin -inkey rsa-key-pub.pem -in helloworld.txt -out helloworld.enc
+// Decrypt file
+$ openssl pkeyutl -decrypt -inkey rsa-key.pem -in helloworld.enc -out helloworld2.txt
+```
+
+If you receive an error, try with `openssl rsautl` instead.
+
+#### Sign and verify
+
+Alice wants to make sure that Bob's public announcement is indeed from Bob:
+
+* Bob gives Alice his public key
+* Bob signs his announcement with his private key
+* Bob sends Alice his announcement and its signature
+* Alice verifies the signature with Bob's public key
+
+![](images/00_12_digital_signature_keys_v2.png)
+
+Example:
+
+```bash
+// Sign file hash
+$ openssl dgst -sha256 -sign secp256k1-key.pem -out helloworld-bin.sha256 helloworld.txt
+// Encode signature in Base64
+$ openssl base64 -in helloworld-bin.sha256 -out helloworld.sha256
+
+// Decode signature form Base64
+$ openssl base64 -d -in helloworld.sha256 -out helloworld-bin-decoded.sha256
+// Verify signature
+$ openssl dgst -sha256 -verify secp256k1-key-pub.pem -signature helloworld-bin-decoded.sha256 helloworld.txt
+Verified OK
+```
+
+#### Mix and match
+
+It is possible to mix both ideas, whereby Alice encrypts her message with Bob's public key, then signs the encrypted file with her private key. Upon reception, Bob verifies the signature with Alice's public key, to be sure it came from Alice, then decrypts the file with his private key.
+
+#### What is this sorcery?
+
+If these examples seem counter-intuitive it means you sense the mathematical wizardry of public key encryption. It's not especially important that you understand the math at a deep level, but it is important that you understand the properties of it. 
+
+Given four keys: A,B,C and D, we can encrypt a message with A, B and C such that D is required to decrypt it and D is very hard to guess or discover. So, if Alice knows her private key and her public key and she also knows Bob's public key, she can encrypt a message that can only be understood by someone with knowledge of Bob's private key. 
+
+Similarly, given knowledge of one's public and private keys, one can generate a signature (a character string) such that someone with a copy of the message and the signature can independely determine the public key of the entity that signed the message and know that the signer has knowledge of the corresponding private key. 
+
+We can proceed with the understanding that signed messages from Alice could only come from someone with knowledge of Alice's private key and messages that are encrypted for Bob can only be deciphered by Bob. 
+
+![](images/00_13_mix_n_match_keys_v3.png)
```

Comment:
Moved the image further up because I feel it would better fit into the content before diving into the recap "What is this sorcery?"